lib/matplotlib/figure.py

```diff
@@ -1576,6 +1576,110 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
             # Returned axis array will be always 2-d, even if nrows=ncols=1.
             return axarr
 
+    def build_grid(self, layout, *, subplot_kw=None, gridspec_kw=None,
+                   empty_sentinel=None):
+        """Build a layout based on an grid of strings.
+
+        This is a helper function to build complex GridSpec layouts visually.
+
+        Parameters
+        ----------
+        layout : list of list of {str, int,  Axes} or nested
+
+            A visual layout of how you want your Axes to be arranged
+            labeled as strings.  For example ::
+
+               x = [['A', 'A', 'B'],
+                    ['C', 'D', 'B']]
+
+            Produces 4 axes:
+
+            - 'A' which is 1 row high and spans the first two columns
+            - 'B' which is 2 rows high and is on the right edge
+            - 'C' which in 1 row and 1 column wide in the bottom left
+            - 'D' which in 1 row and 1 column wide in the bottom center
+
+        subplot_kw : dict, optional
+            Dict with keywords passed to the
+            :meth:`~matplotlib.figure.Figure.add_subplot` call used to create
+            each subplot.
+
+        gridspec_kw : dict, optional
+            Dict with keywords passed to the
+            `~matplotlib.gridspec.GridSpec` constructor used to create
+            the grid the subplots are placed on.
+
+        empty_sentinel : object, optional
+            Entry in the layout to mean "leave this space empty"
+
+
+        Returns
+        -------
+        dict[str, Axes]
+           A dictionary mapping the string labels to the new Axes objects.
+
+        """
+        subplot_kw = subplot_kw or {}
+        gridspec_kw = gridspec_kw or {}
+
+        def _process_layout(layout):
+            unique_ids = set()
+            nested = {}
+            for j, row in enumerate(layout):
+                for k, v in enumerate(row):
+                    if v == empty_sentinel:
+                        continue
+                    elif isinstance(v, (int, str, Axes)):
+                        unique_ids.add(v)
+                    else:
+                        nested[(j, k)] = v
+
+            return unique_ids, nested
+
+        def _do_layout(gs, layout, unique_ids, nested):
+            rows, cols = layout.shape
+            covered = np.zeros((rows, cols), dtype=bool)
+            output = dict()
+
+            for j, k in np.argwhere(layout == empty_sentinel):
+                covered[j, k] = True
+
+            for name in unique_ids:
+                indx = np.argwhere(layout == name)
+                start_row, start_col = np.min(indx, axis=0)
+                end_row, end_col = np.max(indx, axis=0) + 1
+                slc = (slice(start_row, end_row), slice(start_col, end_col))
+                if np.any(covered[slc]):
```

Comment:
I think you can instead check
```
if (layout[slc] != name).any():
    raise ValueError(...)
```
and then the nice thing is that you won't need to keep track of (and thus define) `covered` at all, because the check above is enough to verify that you only have contiguous blocks.  (You need to `continue` first at the top if `name == empty_sentinel`, instead of below.)

Reply:
that is clever.